File: CSES/Mathematics/P1082.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class P1082{

   public static long MOD = 1000000007L;
   public static long i2 = 500000004L;
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      long n = Long.parseLong(f.readLine());
      
      long answer = 0L;
      
      long i = 1L;
      
      while(i <= n){
         long freq = n/i;
         //figure out how many other numbers have the same freq
         long l = i;
         long r = n;
         long ans = i;
         
         while(l <= r){
            long mid = l + (r-l)/2L;
            
            if(n/mid == freq){
               ans = mid;
               l = mid+1;
            } else {
               r = mid-1;
            }
         }
         
         long s1 = (i+r + MOD)%MOD;
         long s2 = (r-i+1 + MOD)%MOD;
         long sum = (s1*s2 + MOD)%MOD;
         long prod = (sum * i2 + MOD)%MOD;
         prod = (prod * freq + MOD)%MOD;
         answer += prod;
         if(answer >= MOD) answer -= MOD;
         
         i = r+1;
      }
      
      out.println(answer);
      
      
      
      
      
      
      
      out.close();
   }
   
      
}

File: CSES/Mathematics/P1082.cpp
#include <bits/stdc++.h>

using namespace std;

const long long MOD = 1000000007LL;
const long long i2 = 500000004LL;

int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   long long n;
   cin >> n;
   
   long long answer = 0LL;
   
   long long i = 1LL;
   
   while(i <= n){
      long long freq = n/i;
      long long l = i;
      long long r = n;
      long long ans = i;
      while(l <= r){
         long long mid = l + (r-l)/2L;
         
         if(n/mid == freq){
            ans = mid;
            l = mid+1;
         } else {
            r = mid-1;
         }
      }
      
   
      long long s1 = (i+r + MOD)%MOD;
      long long s2 = (r-i+1 + MOD)%MOD;
      long long prod = (s1*s2 + MOD)%MOD;
      prod = (prod * i2 + MOD)%MOD;
      prod = (prod * freq + MOD)%MOD;
      answer += prod;
      if(answer >= MOD) answer -= MOD;
      
      i = r+1;
   }
   
   cout << answer << endl;
   
   
   return 0;
}